pageserver/src/walingest.rs

```diff
@@ -107,6 +107,139 @@ struct WarnIngestLag {
     timestamp_invalid_msg_ratelimit: RateLimit,
 }
 
+enum HeapamRecord {
```

Comment:
A high level comment that explains what these structs are for would be nice. Something like:

```
// These structs are an intermediary representation of the PostgreSQL WAL records.
```

Reply:
These structs seem similar to the `XlRelmapUpdate`, `XlHeapDelete` etc. structs in `walrecord.rs`. Are these exactly the same idea? Maybe put these in walrecord.rs, too?

Reply:
> These structs seem similar to the `XlRelmapUpdate`, `XlHeapDelete` etc. structs in walrecord.rs.

They are similar indeed. Think of these new structs as higher level constructs: they have all the context required for the PS to act on them. In some cases that was already true for the `Xl...` types as well and in some cases it was not.

The next PR in the series cleans this up a bit by moving all the `Xl...` types into `postgres_ffi` and all of these "higher level representations" in a new `wal_decoder` module.

Reply:
Heikki, we had a brief out of band chat about this work, but I don't think I mentioned the [RFC](https://github.com/neondatabase/neon/pull/8754). Might be good for some extra context.

Reply:
Added a comment in https://github.com/neondatabase/neon/pull/9472/commits/02fbac3f58495ec650909155743c4b17a87af033